### PR TITLE
Treat empty array as NULL; and fix value relation filter with NULL

### LINKED ISSUE #46366: Constraint "not NULL" is completely ignored for a field with a value relation widget with "allow multiple selection" option
### What is the bug or the crash?

Subject says it all.

### Steps to reproduce the issue

Set up a value relation for a field, check the option "allow multiple selection" and check the "not NULL" constraint. As a result the field will always show the green checkmark, even with no option chosen. Remove the "allow multiple selection" and it will work as expected.

### Versions

3.22.1 / 3.16.14

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### LINKED ISSUE #46369: "current_value" fails to evaluate NULL fields when a form opens in an expression for a value relation filter
### What is the bug or the crash?

Take a form where you have a field that has a value relation/map/text edit, with NULL as default value, call it "category1".

In another field, call it "myfield", you can define an expression for the default value as

`coalesce(current_value('category1'),'foo')`

and when the form opens "myfield" is correctly set with "foo" as default. So far so good.

Now consider another scenario, where another field, call it "myotherfield", is populated with a value relation that must be filtered depending on the value of "category1". In the value relation configs an expression for the filter can be set as

`"category" = coalesce(current_value('category1'),'nonexistingvalue')`

The expected behavior when the forms open would be "myotherfield" be empty because the expression would be evaluated as 

`"category" = 'nonexistingvalue'`

but instead is not filtered at all.

Curiously enough if "myotherfield" is text edit, if its default NULL value is written manually, then "current_value" seems to pick up, but that do not certainly help much

https://user-images.githubusercontent.com/1951107/144742347-a9613c23-49bb-4367-8d99-352d5d78cc23.mp4

### Steps to reproduce the issue

See above.

### Versions

3.16.14 / 3.22.1

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by elpaso**
Fix #46369: 46369 value relation null filtering
Fix #46366: not null constraint ignored when allowmulti

### CODE REVIEW

File: src/core/vector/qgsvectorlayerutils.cpp [hunk 1]
```diff
@@ -431,9 +431,20 @@ bool QgsVectorLayerUtils::validateAttribute( const QgsVectorLayer *layer, const
 
     if ( !exempt )
     {
-      valid = valid && !value.isNull();
 
-      if ( value.isNull() )
+      const bool allowMulti { field.editorWidgetSetup().config().value( QStringLiteral( "AllowMulti" ), false ).toBool() };
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayerutils.cpp:435]: Can we move this into a `bool QgsFieldFormatter::validate()` (or similar) method?

Comment (elpaso) [reply src/core/vector/qgsvectorlayerutils.cpp:435]: maybe, with what kind of signature? The formatter has no knowledge of constraints.

**Comment by gioman** [comment]
@elpaso thanks, thanks... thanks!

File: src/core/vector/qgsvectorlayerutils.cpp (hunk 1 above)
Comment (m-kuhn) [reply src/core/vector/qgsvectorlayerutils.cpp:435]: `isNull()`?

**Comment by m-kuhn** [comment]
I'm half convinced by this approach.

1. Technically, an empty array is not null
This is what postgres tells me:

```sql
select '{}'::int[] is null;
 ?column? 
----------
 f
```

2. QGS Core should not be reading into widget configs. It should call methods on fieldformatter where each widget can deal with its config.
3. The approach taken in here relies on the array being stored in a text field with hstore formatting, it will not work if the underlying data is stored as a native array or json structure in the provider.

**Comment by elpaso** [comment]
> I'm half convinced by this approach.
> 
>     1. Technically, an empty array is not null
>        This is what postgres tells me:
> 
> 
> ```sql
> select '{}'::int[] is null;
>  ?column? 
> ----------
>  f
> ```
> 
>     2. QGS Core should not be reading into widget configs. It should call methods on fieldformatter where each widget can deal with its config.
> 
>     3. The approach taken in here relies on the array being stored in a text field with hstore formatting, it will not work if the underlying data is stored as a native array or json structure in the provider.
> 
> 
> I could also see the following approaches:
> 
>     * Let the widget return NULL for empty arrays, that will work with the unmodified constraint validation logic plus it will match what a database validation does as well
> 
>     * Add a new "must not be empty" setting to the (multi enabled) widget, which will give hints on the gui
> 
>     * Hand this off to the user with a `array_length("field") > 0` function, but that won't work for text fields
> 
>     * Move this logic into a dedicated `bool QgsFieldFormatter::isNull(QVariant)` method, but as mentioned, that actually changes the meaning of null

I would not be much worried about what NULL exactly means for a DB in this context, I think that the user intention when she sets a NOT NULL constraint in these widgets doesn't leave much space to interpretations and can be read as "force the user to make a (multiple?) choice and do not allow a blank".

I will look into the widget returning invalid, it looks the most logical to me, thanks for the hints.

**Comment by elpaso** [comment]
@m-kuhn I think I've found the right approach with my last commit.

Of course from a UX the situation is less than ideal, see the various mentions to `NULL` an `null` in the widget config plus the conflicting options to allow `NULL` in the widget and set not `null` (notice the case) in the constraints section.

Anyway, this is a small fix for the value-relation only, I guess we should check the other widgets as well.

**Review by m-kuhn — APPROVED** [review]
Looks good now, thanks.
Agreed, the null handling (and consistent spelling) could be improved.
Also it would be nice to change the allow null text to allow empty if allow multi is checked.
 Cc @DelazJ

**Comment by DelazJ** [comment]
@m-kuhn sorry for the delay. What was the question, please?

**Comment by m-kuhn** [comment]
I thought the topic of wording on the config UI discussed in https://github.com/qgis/QGIS/pull/46439#issuecomment-991581012 could be interesting for you

**Pull request merged**